### DISCUSSION TITLE
why do you need to explicitly return an HttpHandler when you need the HttpContext?

### DISCUSSION THREAD

**Discussion by Melomaniash** [giraffe-fsharp/Giraffe#475] [Q&A] [answered]
i'm confused as to why the docs state that you have to return an httphandler explicitly when you need the context.

bellow i wrote the function as stated in the docs and then tried writing it directly into the function.
```fs
let addExampleRecord =
        fun (next : HttpFunc) (ctx : HttpContext) ->
            task {
                let! exampleRecord = ctx.BindJsonAsync<ExampleRecord>()
                do! ExampleDataModule.storeExampleRecord exampleRecord

                return! Successful.OK exampleRecord next ctx
            }

let addExampleRecord2 (next : HttpFunc) (ctx : HttpContext) =
    task {
        let! exampleRecord = ctx.BindJsonAsync<ExampleRecord>()
        do! ExampleDataModule.storeExampleRecord exampleRecord

        return! Successful.OK exampleRecord next ctx
    }
```
Both functions worked. So my question is, is there a difference between these 2 functions? and if there is, what is it?

**Comment by dustinmoris** [comment, marked as answer]
Hi, great question and the answer is that there is no difference.

However, there are two reasons which I can think of why some people prefer the first case:

1. You can annotate the return type to be `HttpHandler` and then let the linter and compiler help you to put the curried parameters in the correct order and make sure that the return type matches what a `HttpHandler` expects. It's a facility to help write correct code.
2. If you have more curried arguments before the `next` and `ctx` then you could run some work in advance before the `HttpHandler` part is being returned.

Example:

```fsharp
let addExampleRecord (foo : string) (bar : int) =
        // Do some work only once here using foo and bar
        let fooBar = foo + bar.ToString()
        fun (next : HttpFunc) (ctx : HttpContext) ->
            task {
                let! exampleRecord = ctx.BindJsonAsync<ExampleRecord>()
                do! ExampleDataModule.storeExampleRecord exampleRecord
                // Use fooBar here...
                return! Successful.OK exampleRecord next ctx
            }
```
